src/libraries/System.Net.Requests/src/System/Net/RequestBufferingStream.cs

```diff
@@ -0,0 +1,134 @@
+        public override bool CanWrite => true;
+
+        public override void Flush() => ThrowIfDisposed(); // Nothing to do.
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+            // Nothing to do.
+            return cancellationToken.IsCancellationRequested ?
+                Task.FromCanceled(cancellationToken) :
+                Task.CompletedTask;
+        }
+
+        public override long Length
+        {
+            get
+            {
+                throw new NotSupportedException();
+            }
+        }
+
+        public override long Position
+        {
+            get
+            {
+                throw new NotSupportedException();
+            }
+            set
+            {
+                throw new NotSupportedException();
+            }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            ThrowIfDisposed();
+            ValidateBufferArguments(buffer, offset, count);
+            _buffer.Write(buffer, offset, count);
+        }
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+            ValidateBufferArguments(buffer, offset, count);
+            return _buffer.WriteAsync(buffer, offset, count, cancellationToken);
+        }
+
+        public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            ThrowIfDisposed();
+            return _buffer.WriteAsync(buffer, cancellationToken);
+        }
+
+        public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback? asyncCallback, object? asyncState)
+        {
+            ThrowIfDisposed();
+            ValidateBufferArguments(buffer, offset, count);
+            return _buffer.BeginWrite(buffer, offset, count, asyncCallback, asyncState);
+        }
+
+        public override void EndWrite(IAsyncResult asyncResult)
+        {
+            ThrowIfDisposed();
+            _buffer.EndWrite(asyncResult);
+        }
+
+        public ArraySegment<byte> GetBuffer()
+        {
+            ArraySegment<byte> bytes;
+
+            bool success = _buffer.TryGetBuffer(out bytes);
+            Debug.Assert(success); // Buffer should always be visible since default MemoryStream constructor was used.
+
+            return bytes;
+        }
+
+        // We need this to dispose the MemoryStream.
```

Comment:
Why doesn't this class `Dispose` dispose the `MemoryStream`?

Reply:
Because user will be controlling this class and they can choose to dispose it before we send the request (Before calling `GetResponse`), in that case we'll try to use disposed `MemoryStream`

Reply:
> they can choose to dispose it before we send the request (Before calling GetResponse)

Is this a valid use that is expected to work? Are you saying that this sequence of actions work on Framework?

Reply:
Yes, it works on .NET Framework

```c#
HttpWebRequest request = (HttpWebRequest)WebRequest.Create("http://httpbin.org/post");
request.Method = "POST";
using (var requestStream = request.GetRequestStream())
{
    byte[] data = System.Text.Encoding.UTF8.GetBytes("Hello World");
    requestStream.Write(data, 0, data.Length);
}
using (var response = request.GetResponse())
{
    using (var responseStream = response.GetResponseStream())
    {
        using (var reader = new System.IO.StreamReader(responseStream))
        {
            string responseFromServer = reader.ReadToEnd();
            Console.WriteLine(responseFromServer);
        }
    }
}
Console.ReadLine();
```